### ISSUE TITLE
Collect source-paths from project-specs

### ISSUE THREAD

**Issue by bendlas** [clojure-lsp/clojure-lsp#551]
**Is your feature request related to a problem? Please describe.**

When working with a code base consisting of multiple projects, I have to add `:source-paths` from all submodules, in order to get static analysis to work.

Otherwise, only such files are analyzed, that I already opened within the lsp session.

**Describe the solution you'd like**

Intuitively, I've tried adding all the submodules into `:project-specs`.

**Describe alternatives you've considered**

@ericdallo [proposed](https://github.com/clojure-lsp/clojure-lsp/issues/191#issuecomment-918391332) adding support for `lein-monolith` specifically.

I think support for individual leiningen plugins may be worthwhile, however, there are many different ones that are also being used instead of `lein-monolith`, including project types other than leiningen. They all use different configuration formats so before adding translation layers for any of them, `clojure-lsp` should have a clearly documented support for any sort of subproject structure.

- https://github.com/kumarshantanu/lein-sub
- https://github.com/jcrossley3/lein-modules
- https://github.com/amperity/lein-monolith
- ...

**Additional context**

Here is the [snippet, I currently use for generating my `:source-paths`](https://gist.github.com/bendlas/4a2a98b1cbe1d1efa8a7ef6850934e13).

**Comment by ericdallo** [comment]
Thank you for the issue!

We could add support for lein-monolith but as you said, adding support for each of multiple plugins may be too much, maybe we can take an approach like you snippet to generate source-paths for whatever lein project, one of the issues I see is the need of leiningen context/requires to use that snippet

**Comment by bendlas** [comment]
Wouldn't that also entail graal-compiling leiningen? Not sure if that's an option, TBH.

It would seem natural to re-use the existing [source path collection](https://github.com/clojure-lsp/clojure-lsp/blob/8a6cdb90fa2d0835bf74ddc005ac3e9b14db9424/src/clojure_lsp/source_paths.clj#L73). Apparently, that already does the `:profiles` expansion, that I also do in my snippet (y).

With that, the missing link would be how to collect subproject roots from a parent project, but having a config option for additional source-path roots should make such enhancements significantly easier.

So, `:project-specs` is for dependency collection, right? Is that also the right option to overload for source-path collection?

**Comment by ericdallo** [comment]
> So, :project-specs is for dependency collection, right? Is that also the right option to overload for source-path collection?

No, that is just for bringing external deps for clj-kondo analysis, so the source-paths discovery is something different unfortunately.

> It would seem natural to re-use the existing source path collection. Apparently, that already does the :profiles expansion, that I also do in my snippet (y).
With that, the missing link would be how to collect subproject roots from a parent project, but having a config option for additional source-path roots should make such enhancements significantly easier.

Yeah, we already slurp and static analyze the project.clj, but we would need to somehow know what are the subprojects and merge the source-paths from those subprojects.

**Comment by bendlas** [comment]
> No, that is just for bringing external deps for clj-kondo analysis, so the source-paths discovery is something different unfortunately.

Oh, in that case: shall I rewrite the ticket in terms of a new configuration option `:subproject-roots`?

**Comment by ericdallo** [comment]
@bendlas I don't think we need a `:subproject-roots`, you can already use `:source-paths #{"subproject-1" "subproject-2"}` right?
My point is that I would like to improve clojure-lsp to automatically discover those subprojects

**Comment by bendlas** [comment]
Ah yes, auto-discovery is another option to consider as well. I'd tend to not start there, and instead add it as sugar, or option default.

> @bendlas I don't think we need a `:subproject-roots`, you can already use `:source-paths #{"subproject-1" "subproject-2"}` right?

Overloading `:source-paths` for this seems dangerous. What if I have a file called `deps.edn` or `project.clj` as a regular source file in my classpath root?

**Comment by bendlas** [comment]
> @bendlas I don't think we need a `:subproject-roots`, you can already use `:source-paths #{"subproject-1" "subproject-2"}` right?

Actually, let me change that proposal to `:local/roots` or `:local-roots`, because `deps.edn` already has a well-defined source for this, called `:local/root`.

**Comment by ericdallo** [comment]
I see, so it'd be a different flag that will use the same logic it already uses to detect source-paths on project-root but for those specified dirs, right?
Do you think that would work for most lein plugins? 

If so, clojure-lsp would merge the source-path discovery logic applied to those subdirectories + the project root one

**Comment by bendlas** [comment]
> I see, so it'd be a different flag that will use the same logic it already uses to detect source-paths on project-root but for those specified dirs, right?

Yes, essentially that.

> Do you think that would work for most lein plugins?

In the sense that it would provide a clear target for supporting them, as part of further enhancements.

Unfortunately, leiningen doesn't have any sort of common infrastructure around that, AFAIK.

**Comment by ericdallo** [comment]
Alright, I think it's a valid idea add a `:source-subpaths` as a set of relative paths, we will need to test if it's enough for lein-monolilth and other plugins though

**Comment by bendlas** [comment]
After I wrongfully assumed that `:project-specs`, I think we also need dependency collection, i.e. to apply [](https://github.com/clojure-lsp/clojure-lsp/blob/234baa277634e738bdc7214cd2349611792ab7a8/src/clojure_lsp/crawler.clj#L68) to a set of relative paths in a similar fashion.

I can't think of a case where you'd collect source paths from a relative path, but not dependencies, but if you think it preferrable, that can be a separate option `:dependency-subpaths` or similar.

**Comment by ericdallo** [comment]
for `:project-specs`, clojure-lsp will end up running a `lein classpath` in the project-root, won't that will bring the subprojects?

**Comment by ericdallo** [comment]
for `:project-specs`, clojure-lsp will end up running a `lein classpath` in the project-root, won't that will bring the subprojects? 

Also, the source-paths and classpath discovery are already too complex, I'd like to avoid adding more external vars complexity, that's why I suggested the automatic one

**Comment by ericdallo** [comment]
Just giving a lot of more context about those things on clojure-lsp:
- clojure-lsp needs to know the classpath/files to analyze with clj-kondo to most of the features work, this is known as the classpath scan which is done during startup, the var to customize that is `:project-specs`.
- clojure-lsp also needs to know the source-paths of the project, to features like rename, auto add new ns forms for new files and other features, the var to customize that is `:source-paths` and `:source-aliases`. clojure-lsp also scan the source-paths for analysis.

I think we can start improving the source-path discoverability for lein mono repos as this is necessary for clojure-lsp features and it will give analysis as well

**Comment by bendlas** [comment]
> for `:project-specs`, clojure-lsp will end up running a `lein classpath` in the project-root, won't that will bring the subprojects?

Only if the plugin hooks into the classpath, to add the subprojects. Many don't; including the use case of not having a root `project.clj` at all, but just a buch of checkouts of interacting projects.

> * clojure-lsp needs to know the classpath/files to analyze with clj-kondo to most of the features work, this is known as the classpath scan which is done during startup, the var to customize that is `:project-specs`.

ok, does that also govern the case where `lsp-find-definition` jumps into a file from the classpath?

**Comment by ericdallo** [comment]
> Only if the plugin hooks into the classpath, to add the subprojects. Many don't; including the use case of not having a root project.clj at all, but just a bunch of checkouts of interacting projects.

Yeah, this sounds a bad pattern have code that is not included on the classpath in the end, anyway, adding that new setting show fix those cases right?

> ok, does that also govern the case where lsp-find-definition jumps into a file from the classpath?

Yes, as clojure-lsp needs to have the analysis for that file to know that the definition is there.

**Comment by bendlas** [comment]
> Yeah, this sounds a bad pattern have code that is not included on the classpath in the end

I guess it is.

> adding that new setting show fix those cases right?
> 
> > ok, does that also govern the case where lsp-find-definition jumps into a file from the classpath?
> 
> Yes, as clojure-lsp needs to have the analysis for that file to know that the definition is there.

Sorry, I'm a bit confused now. So the classpath can be customized by `:project-specs`, will be analyzed for the purposes of `find-definition`, but ignored for `:source-paths` so that lsp won't try to rewrite stuff in your `.m2` folder; is that correct so far?

If so, I can add a project folder to `:source-subpaths` and `:project-specs`, in order to get both source folders and dependencies from it, right?

**Comment by ericdallo** [comment]
clojure-lsp during startup it:
- analyzes the classpath, `lein classpath`, and then caches it, only analyzing in a next deps change or not presence of `.lsp/.cache`.
- **Always** analyzes the project source-paths

> so that lsp won't try to rewrite stuff in your .m2 folder

sorry, I don't get it what you mean with that, clojure-lsp doesn't write anything to .m2, only analyze the paths from the return of lein classpath which probably has files on .m2

**Comment by ericdallo** [comment]
@bendlas Made a commit on master using source-paths from classpath, check #752, could you confirm it works for you?

**Comment by bendlas** [comment]
@ericdallo I attempted to run that fix in my NixOS configuration. Since there is no released uberjar for this (right?), I modified [the build recipe](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/misc/clojure-lsp/default.nix#L14) to build its own `clojure-lsp-standalone.jar` (based on clojure-lsp commit ca1621ba7f8475095f2bf74d895411372debf4f4).

I've tried two different methods to build the uberjar, (`make cli-jar-for-native`, `make cli-jar`) but both make the build fail at different stages. I was wondering if I'm building the uberjars wrong, or if something else changed since `2022.02.01-16.53.14`.

## `make cli-jar-for-native`

When I build the jar with this, the build fails at step 2:

```
[2/7] Performing analysis...  [*]                                                                       (87.2s @ 2.17GB)
To see how the classes got initialized, use --trace-class-initialization=clj_easy.graal_build_time$packages_from_jar,clj_easy.graal_build_time$_packageListStr,clj_easy.graal_build_time$unique_packages,clj_easy.graal_build_time$_packageList,clj_easy.graal_build_time$packages_from_dir,clj_easy.graal_build_time$entry__GT_package,clj_easy.graal_build_time$consider_entry_QMARK_,clj_easy.graal_build_time$contains_parent_QMARK_
   6,840 (87.52%) of  7,815 classes reachable
   8,192 (49.96%) of 16,398 fields reachable
  30,673 (64.01%) of 47,918 methods reachable
     426 classes,   807 fields, and 2,277 methods registered for reflection

Error: Classes that should be initialized at run time got initialized during image building:
 clj_easy.graal_build_time$packages_from_jar was unintentionally initialized at build time. To see why clj_easy.graal_build_time$packages_from_jar got initialized use --trace-class-initialization=clj_easy.graal_build_time$packages_from_jar
clj_easy.graal_build_time$_packageListStr was unintentionally initialized at build time. To see why clj_easy.graal_build_time$_packageListStr got initialized use --trace-class-initialization=clj_easy.graal_build_time$_packageListStr
clj_easy.graal_build_time$unique_packages was unintentionally initialized at build time. To see why clj_easy.graal_build_time$unique_packages got initialized use --trace-class-initialization=clj_easy.graal_build_time$unique_packages
clj_easy.graal_build_time$_packageList was unintentionally initialized at build time. To see why clj_easy.graal_build_time$_packageList
got initialized use --trace-class-initialization=clj_easy.graal_build_time$_packageList
clj_easy.graal_build_time$packages_from_dir was unintentionally initialized at build time. To see why clj_easy.graal_build_time$packages_from_dir got initialized use --trace-class-initialization=clj_easy.graal_build_time$packages_from_dir
clj_easy.graal_build_time$entry__GT_package was unintentionally initialized at build time. To see why clj_easy.graal_build_time$entry__GT_package got initialized use --trace-class-initialization=clj_easy.graal_build_time$entry__GT_package
clj_easy.graal_build_time$consider_entry_QMARK_ was unintentionally initialized at build time. To see why clj_easy.graal_build_time$consider_entry_QMARK_ got initialized use --trace-class-initialization=clj_easy.graal_build_time$consider_entry_QMARK_
clj_easy.graal_build_time$contains_parent_QMARK_ was unintentionally initialized at build time. To see why clj_easy.graal_build_time$contains_parent_QMARK_ got initialized use --trace-class-initialization=clj_easy.graal_build_time$contains_parent_QMARK_
```

## `make cli-jar`

This actually completes step 7, but the tests fail .. not so much of a surprise, since it's built without the native profile.

```
Finished generating 'clojure-lsp' in 3m 58s.
running tests
Exception in thread "main" java.lang.ExceptionInInitializerError
        at clojure.lang.Namespace.<init>(Namespace.java:34)
        at clojure.lang.Namespace.findOrCreate(Namespace.java:176)
        at clojure.lang.Var.internPrivate(Var.java:156)
        at clojure_lsp.main.<clinit>(Unknown Source)
Caused by: java.io.FileNotFoundException: Could not locate clojure/core__init.class, clojure/core.clj or clojure/core.cljc on classpath.
        at clojure.lang.RT.load(RT.java:462)
        at clojure.lang.RT.load(RT.java:424)
        at clojure.lang.RT.<clinit>(RT.java:338)
        ... 4 more
```

## The jarfile recipe

```nix
  jar = stdenv.mkDerivation {
    pname = "clojure-lsp-standalone.jar";
    inherit src version;
    buildInputs = [ clojure git cacert ];
    preBuild = ''
      export HOME=`pwd`/homedir
      mkdir -p $HOME
    '';
    makeFlags = [ "cli-jar-for-native" ];
    # makeFlags = [ "cli-jar" ];
    installPhase = ''
      cp clojure-lsp-standalone.jar $out
    '';
    outputHashAlgo = "sha256";
    outputHash = "sha256-6YEGzOrodR5bSX0PsfDuy3qcD6bYTGZlo8Lb6YEOwk0=";
    # outputHash = "sha256-9d3kLl1XUls5puE/KC1VqN45kePTJgtRuNgw5hFclZY=";
    outputHashMode = "flat";
  };
```

**Comment by ericdallo** [comment]
@bendlas I'm a NixOS user as well, I think the fastest way to test that is download a nightly build from [#clojure-lsp-builds](https://clojurians.slack.com/archives/C032YH7P0R2), for long term, #604 should help in the future

**Comment by bendlas** [comment]
Thanks!
The static build doesn't work either, because it still tries to use `/lib64/ld-linux-x86-64.so.2`

```
$ strace /tmp/clojure-lsp/clojure-lsp
execve("/tmp/clojure-lsp/clojure-lsp", ["/tmp/clojure-lsp/clojure-lsp"], 0x7ffdb0d850c0 /* 83 vars */) = 0
open("/proc/self/exe", O_RDONLY)        = 3
mmap(NULL, 27809070, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fdbf9014000
mmap(0x7fdbf9014000, 27808645, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 3, 0) = 0x7fdbf9014000
mprotect(0x7fdbfaa98000, 5422, PROT_READ|PROT_EXEC) = 0
readlink("/proc/self/exe", "/tmp/clojure-lsp/clojure-lsp", 4095) = 28
mmap(0x7fdbfaa9a000, 130519040, PROT_NONE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fdbfaa9a000
mmap(0x7fdbfaa9a000, 23624048, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fdbfaa9a000
mprotect(0x7fdbfaa9a000, 23624048, PROT_READ) = 0
mmap(0x7fdbfc122000, 39863333, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0x1688000) = 0x7fdbfc122000
mprotect(0x7fdbfc122000, 39863333, PROT_READ|PROT_EXEC) = 0
mmap(0x7fdbfe727000, 66987308, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0x3c8d000) = 0x7fdbfe727000
mprotect(0x7fdbfe727000, 66987308, PROT_READ) = 0
mmap(0x7fdc0270a000, 33672, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0x7c6f000) = 0x7fdc0270a000
mprotect(0x7fdc0270a000, 33672, PROT_READ|PROT_WRITE) = 0
open("/lib64/ld-linux-x86-64.so.2", O_RDONLY) = -1 ENOENT (No such file or directory)
exit(127)                               = ?
+++ exited with 127 +++
```
I'll attempt to make use of the flake PR.

**Comment by ericdallo** [comment]
I'm closing this issue as we implemented #752, LMK if any issues
